### ISSUE TITLE
InfStepRange is not an ordinal range

### ISSUE THREAD

**Issue by jishnub** [JuliaArrays/InfiniteArrays.jl#109]
See https://github.com/JuliaArrays/FillArrays.jl/issues/207

Currently, `InfStepRange(` is too permissive:
```julia
julia> InfiniteArrays.InfStepRange(3, 2.5)
3:2.5:+∞
```
If this is an ordinal range, the step size should be limited to multiples of `oneunit`. Perhaps the current design is accidental, rather than the intention, and the start and step should be integers?

**Comment by putianyi889** [comment]
I'm trying to change the definition to 
```julia
struct InfStepRange{T,S} <: AbstractRange{T}
    start::T
    step::S
end
```
Hopefully additional small fixes could pass all tests.

**Comment by jishnub** [comment]
Perhaps this should be an `OrdinalRange` after all, with checks on the step? This will be analogous to a `StepRange`.

**Comment by putianyi889** [comment]
For float input, as for how julia base works, it should return a InfStepRangeLen then. Julia base never generates a StepRange of float type.

**Comment by jishnub** [comment]
Yes, that might make sense for the colon constructor. Ideally, the infinite types should shadow the finite ones

**Comment by putianyi889** [comment]
The motivation of this issue is
```julia
julia> 1.0:∞
1.0:1.0:+∞

julia> float(1:∞)
1.0:∞
```

**Comment by putianyi889** [comment]
well maybe InfiniteArrays is correct while Julia base is not

**Comment by jishnub** [comment]
I don't think the types are considered a part of correctness. The resulting values are the same in each case, with the types in `Base` being chosen to ensure that the values are accurately represented. Something similar might be possible in this package, but the focus has perhaps not been on floating-point ranges as much as integer ones.

**Comment by dlfivefifty** [comment]
Infinite float ranges are definitely important, as operators are often float valued (think of even the Derivative operator for Chebyshev but on a mapped interval)
